src/dvc_data/index.py

```diff
@@ -238,8 +243,11 @@ def collect(index, path, fs):
 def save(index):
     from .hashfile.build import _build_file
 
-    for _, entry in index.iteritems():
+    dir_entries: List[Tuple[DataIndexKey, DataIndexEntry]] = []
+
+    for key, entry in index.iteritems():
         if entry.meta.isdir:
+            dir_entries.append((key, entry))
```

Comment:
dir trees need to be built+saved last, after we've saved all the files that go inside the tree